File: algorithms/03_cpp/data_structures/fenwick_tree.cpp
// Fenwick (binary indexed) tree for prefix sums.
#include <cassert>
#include <iostream>
#include <vector>

class FenwickTree {
  public:
    explicit FenwickTree(int n) : tree_(n + 1, 0) {}
    void add(int index, int delta) {
        for (++index; index < static_cast<int>(tree_.size()); index += index & -index)
            tree_[index] += delta;
    }
    int prefix_sum(int index) {
        int total = 0;
        for (++index; index > 0; index -= index & -index) total += tree_[index];
        return total;
    }
    int range_sum(int left, int right) { return prefix_sum(right) - prefix_sum(left - 1); }

  private:
    std::vector<int> tree_;
};

int main() {
    FenwickTree tree(8);
    int values[] = {1, 3, 5, 7, 9, 11};
    for (int i = 0; i < 6; ++i) tree.add(i, values[i]);
    assert(tree.prefix_sum(3) == 16);
    assert(tree.range_sum(2, 4) == 21);
    std::cout << "fenwick tree ok\n";
    return 0;
}

File: algorithms/07_typescript/data_structures/fenwick_tree.ts
/**
 * Fenwick Tree / Binary Indexed Tree (Peter Fenwick)
 * O(log n) point update and prefix sum query data structure.
 */

export class FenwickTree {
  private tree: number[];

  constructor(size: number) {
    this.tree = new Array(size + 1).fill(0);
  }

  update(index: number, delta: number): void {
    for (let i = index + 1; i < this.tree.length; i += i & -i) {
      this.tree[i] += delta;
    }
  }

  prefixSum(index: number): number {
    let sum = 0;
    for (let i = index + 1; i > 0; i -= i & -i) {
      sum += this.tree[i];
    }
    return sum;
  }

  rangeSum(left: number, right: number): number {
    return this.prefixSum(right) - (left > 0 ? this.prefixSum(left - 1) : 0);
  }
}

const bit = new FenwickTree(10);
[1, 3, 5, 7, 9, 11].forEach((v, i) => bit.update(i, v));
if (bit.rangeSum(1, 3) !== 15) throw new Error("Fenwick tree range sum error");
console.log("Fenwick Tree verified successfully.");
